File: editor-server/src/resolvers/led.ts
import {
  Resolver,
  Query,
  Arg,
  Ctx,
  Mutation,
  PubSub,
  Publisher,
  Int,
} from "type-graphql";

import { LEDMap } from "./types/ledEffectMap";
import { DeleteLEDInput } from "./inputs/led";
import {
  LEDEffectResponse,
  DeleteLEDEffectResponse,
} from "./response/ledEffectResponse";
import { EditLEDInput } from "./inputs/led";
import { TContext } from "../types/global";
import { LEDEffectCreateInput } from "../../prisma/generated/type-graphql";
import { Topic } from "./subscriptions/topic";
import { LEDPayload, ledMutation } from "./subscriptions/led";

@Resolver()
export class LEDResolver {
  @Query((returns) => LEDMap)
  async LEDMap(@Ctx() ctx: TContext) {
    const allPart = await ctx.prisma.part.findMany({ where: { type: "LED" } });
    return { LEDMap: allPart };
  }

  @Mutation((returns) => LEDEffectResponse)
  async addLEDEffect(
    @PubSub(Topic.LEDRecord)
    publishLEDRecord: Publisher<LEDPayload>,
    @Arg("input") input: LEDEffectCreateInput,
    @Ctx() ctx: TContext
  ) {
    const { name, partName, repeat, frames } = input;

    // check part validity
    const part = await ctx.prisma.part.findMany({
      where: { name: partName, type: "LED" },
    });
    if (part.length === 0) {
      return Object.assign({
        id: -1,
        partName: partName,
        effectName: name,
        repeat: repeat,
        effects: [],
        ok: false,
        msg: "no corresponding part.",
      });
    }

    // check if the color in LED is valid
    const error: string[] = [];
    const colors = await ctx.prisma.color.findMany({ select: { id: true } });
    const colorIds = colors.map((color) => color.id);
    if (frames) {
      frames.set.map((frame: any) => {
        frame.LEDs.map((LED: number[]) => {
          if (LED.length !== 2) {
            error.push(`LEDs shape invalid.`);
            return;
          }
          if (!colorIds.includes(LED[0])) {
            error.push(`Color Id ${LED[0]} not found.`);
          }
        });
      });
      if (error.length !== 0) {
        return Object.assign({
          id: -1,
          partName: ",",
          effectName: ",",
          repeat: 0,
          effects: [],
          ok: false,
          msg: error.join(" | "),
        });
      }
    }

    // check overlapped
    const exist = await ctx.prisma.lEDEffect.findMany({
      where: { name: name, partName: partName },
    });
    if (exist.length !== 0)
      return Object.assign({
        id: -1,
        partName: partName,
        effectName: name,
        effects: [],
        repeat: repeat,
        ok: false,
        msg: "effectName exist.",
      });

    const newLED = await ctx.prisma.lEDEffect.create({
      data: {
        name: name,
        partName: partName,
        repeat: repeat,
        frames: frames?.set,
      },
    });

    const recordPayload: LEDPayload = {
      mutation: ledMutation.CREATED,
      editBy: ctx.userId,
      id: newLED.id,
      partName,
      effectName: name,
      data: newLED,
    };
    await publishLEDRecord(recordPayload);

    return Object.assign({
      id: newLED.id,
      partName: partName,
      effectName: name,
      repeat: repeat,
      effects: frames?.set,
      ok: true,
      msg: "successfully add LED effect",
    });
  }

  @Mutation((returns) => LEDEffectResponse)
  async editLEDEffect(
    @PubSub(Topic.LEDRecord)
    publishLEDRecord: Publisher<LEDPayload>,
    @Arg("input") input: EditLEDInput,
    @Ctx() ctx: TContext
  ) {
    const { id, name, repeat, frames } = input;

    // check overlapped
    const exist = await ctx.prisma.lEDEffect.findFirst({
      where: { id: id },
    });
    if (!exist)
      return Object.assign({
        id: -1,
        partName: ",",
        effectName: ",",
        repeat: 0,
        effects: [],
        ok: false,
        msg: "effectName do not exist.",
      });
    const effectToEdit = await ctx.prisma.editingLEDEffect.findFirst({
      where: { LEDEffectId: id },
    });
    if (
      effectToEdit &&
      effectToEdit.userId &&
      effectToEdit.userId !== ctx.userId
    ) {
      throw new Error(`The frame is now editing by ${effectToEdit.userId}.`);
    }

    // check if the color in LED is valid
    const error: string[] = [];
    const colors = await ctx.prisma.color.findMany({ select: { id: true } });
    const colorIds = colors.map((color) => color.id);
    if (frames) {
      frames.set.map((frame: any) => {
        frame.LEDs.map((LED: number[]) => {
          if (LED.length !== 2) {
            error.push(`LEDs shape invalid.`);
            return;
          }
          if (!colorIds.includes(LED[0])) {
            error.push(`Color Id ${LED[0]} not found.`);
          }
        });
      });
      if (error.length !== 0) {
        return Object.assign({
          id: -1,
          partName: ",",
          effectName: ",",
          repeat: 0,
          effects: [],
          ok: false,
          msg: error.join(" | "),
        });
      }
    }

    const target = await ctx.prisma.lEDEffect.update({
      where: {
        id: id,
      },
      data: {
        // add update name
        name: name,
        repeat: repeat,
        frames: frames?.set,
      },
    });

    const recordPayload: LEDPayload = {
      mutation: ledMutation.UPDATED,
      id: target.id,
      editBy: ctx.userId,
      partName: exist.partName,
      effectName: name,
      data: target,
    };
    await publishLEDRecord(recordPayload);

    return Object.assign({
      id: id,
      ok: true,
      msg: "successfully edit LED effect",
      partName: exist.partName,
      effectName: name,
      repeat: target.repeat,
      effects: target.frames,
    });
  }

  @Mutation((returns) => DeleteLEDEffectResponse)
  async deleteLEDEffect(
    @PubSub(Topic.LEDRecord) publishLEDRecord: Publisher<LEDPayload>,
    @Arg("id", () => Int) id: number,
    @Ctx() ctx: TContext
  ) {
    // check exist
    const exist = await ctx.prisma.lEDEffect.findUnique({
      where: { id },
    });
    if (!exist)
      return {
        ok: false,
        msg: `LEDEffect Id ${id} not found`,
      };

    const checkControl = await ctx.prisma.controlData.findMany({
      where: { value: { path: ["src"], equals: id } },
    });
    if (checkControl.length != 0) {
      let checkControlFrames: number[] = checkControl.map(
        (control) => control.frameId
      );
      checkControlFrames = checkControlFrames.sort(function (a, b) {
        return a - b;
      });
      let frame = 0;
      const ids: number[] = [];
      checkControlFrames.map((controlFrame) => {
        if (controlFrame !== frame) {
          ids.push(controlFrame);
          frame = controlFrame;
        }
      });
      return {
        ok: false,
        msg: `LEDeffect Id ${id} is used in frame Ids ${ids}`,
      };
    }
    const effectToDelete = await ctx.prisma.editingLEDEffect.findFirst({
      where: { LEDEffectId: exist.id },
    });
    if (
      effectToDelete &&
      effectToDelete.userId &&
      effectToDelete.userId !== ctx.userId
    ) {
      return {
        ok: false,
        msg: `The frame is now editing by ${effectToDelete.userId}.`,
      };
    }
    await ctx.prisma.lEDEffect.delete({
      where: { id },
    });
    const recordPayload: LEDPayload = {
      mutation: ledMutation.DELETED,
      editBy: ctx.userId,
      id: exist.id,
    };
    await publishLEDRecord(recordPayload);
    return { ok: true, msg: "successfully delete LED effect" };
  }
}

File: editor-server/src/graphql/mutations/led.rs
use crate::graphql::types::led::{Frame, LEDEffectData, LEDEffectFrame};
use crate::graphql::{subscriptions::led::LEDPayload, subscriptor::Subscriptor};
use crate::types::global::UserContext;
use async_graphql::{
    Context, Error as GQLError, InputObject, Object, Result as GQLResult, SimpleObject,
};
use serde::{Deserialize, Serialize};

// struct defined to to fit old schema
#[derive(InputObject, Default, Debug)]
#[graphql(name = "LEDEffectCreateInput")]
pub struct LEDEffectCreateInput {
    pub name: String,
    pub model_name: String,
    pub part_name: String,
    pub repeat: i32,
    pub frames: Vec<Frame>,
}

#[derive(InputObject, Default, Debug)]
#[graphql(name = "EditLEDInput")]
pub struct EditLEDInput {
    pub id: i32,
    pub name: String,
    pub repeat: i32,
    pub frames: Vec<Frame>,
}

#[derive(SimpleObject, Serialize, Deserialize, Default, Debug)]
#[graphql(name = "LEDEffectResponse")]
pub struct LEDEffectResponse {
    id: i32,
    model_name: String,
    part_name: String,
    effect_name: String,
    repeat: i32,
    effects: Vec<LEDEffectFrame>,
    ok: bool,
    msg: String,
}

#[derive(SimpleObject, Serialize, Deserialize, Default, Debug)]
#[graphql(name = "DeleteLEDEffectResponse")]
pub struct DeleteLEDEffectResponse {
    ok: bool,
    msg: String,
}

#[derive(Default)]
pub struct LEDMutation;

#[Object]
impl LEDMutation {
    #[graphql(name = "addLEDEffect")]
    async fn add_led_effect(
        &self,
        ctx: &Context<'_>,
        input: LEDEffectCreateInput,
    ) -> GQLResult<LEDEffectResponse> {
        let context = ctx.data::<UserContext>()?;
        let clients = context.clients;

        let mysql = clients.mysql_pool();

        let effect_name = input.name.clone();
        let model_name = input.model_name.clone();
        let part_name = input.part_name.clone();
        let repeat = input.repeat;
        let frames: Vec<LEDEffectFrame> = input
            .frames
            .iter()
            .map(|frame| LEDEffectFrame {
                leds: frame.leds.clone(),
                fade: frame.fade,
                start: frame.start,
            })
            .collect();

        // check part exists
        let (model_id, part_id) = match sqlx::query!(
            r#"
                SELECT
                    Model.id as "model_id",
                    Part.id as "part_id"
                FROM Part
                INNER JOIN Model ON Part.model_id = Model.id
                WHERE Part.name = ? AND Model.name = ?;
            "#,
            &part_name,
            &model_name,
        )
        .fetch_one(mysql)
        .await
        {
            Ok(row) => (row.model_id, row.part_id),
            Err(_) => {
                return Ok(LEDEffectResponse {
                    id: -1,
                    model_name,
                    part_name,
                    effect_name,
                    repeat,
                    effects: vec![],
                    ok: false,
                    msg: "No corresponding dancer-part pair.".to_string(),
                })
            }
        };

        // create array of all color ids from db
        let colors = sqlx::query!(
            r#"
                SELECT id FROM Color;
            "#
        )
        .fetch_all(mysql)
        .await?;

        // check if all color ids in frames are in db
        for frame in frames.iter() {
            for led in frame.leds.iter() {
                let mut found = false;
                for color in colors.iter() {
                    if led[0] == color.id {
                        found = true;
                        break;
                    }
                }
                if !found {
                    return Ok(LEDEffectResponse {
                        id: -1,
                        model_name,
                        part_name,
                        effect_name,
                        repeat,
                        effects: vec![],
                        ok: false,
                        msg: format!("Color Id {} not found", led[0]),
                    });
                }
            }
        }

        // check if effect name exists
        let effect_id = match sqlx::query!(
            r#"
                SELECT * from LEDEffect
                WHERE name = ? AND part_id = ? AND model_id = ?;
            "#,
            &effect_name,
            part_id,
            model_id
        )
        .fetch_one(mysql)
        .await
        {
            Ok(_) => {
                return Ok(LEDEffectResponse {
                    id: -1,
                    model_name,
                    part_name,
                    effect_name,
                    repeat,
                    effects: vec![],
                    ok: false,
                    msg: "effectName exists.".to_string(),
                })
            }
            Err(_) => sqlx::query!(
                r#"
                    INSERT INTO LEDEffect (name, model_id, part_id)
                    VALUES (?, ?, ?)
                "#,
                &effect_name,
                model_id,
                part_id
            )
            .execute(mysql)
            .await?
            .last_insert_id() as i32,
        };

        // insert frames into LEDEffectStates
        for frame in frames.iter() {
            for (i, led) in frame.leds.iter().enumerate() {
                let _ = sqlx::query!(
                    r#"
                        INSERT INTO LEDEffectState (effect_id, position, color_id, alpha)
                        VALUES (?, ?, ?, ?)
                    "#,
                    effect_id,
                    i as i32,
                    led[0],
                    led[1],
                )
                .execute(mysql)
                .await?;
            }
        }

        // publish to subscribers
        let led_payload = LEDPayload {
            create_effects: vec![LEDEffectData {
                id: effect_id,
                name: effect_name.clone(),
                model_name: model_name.clone(),
                part_name: part_name.clone(),
                repeat,
                frames: frames.clone(),
            }],
            update_effects: Vec::new(),
            delete_effects: Vec::new(),
        };

        Subscriptor::publish(led_payload);

        Ok(LEDEffectResponse {
            id: effect_id,
            model_name,
            part_name,
            effect_name,
            repeat: 0,
            effects: frames,
            ok: true,
            msg: "successfully added LED effect".to_string(),
        })
    }

    #[graphql(name = "editLEDEffect")]
    async fn edit_led_effect(
        &self,
        ctx: &Context<'_>,
        input: EditLEDInput,
    ) -> GQLResult<LEDEffectResponse> {
        let context = ctx.data::<UserContext>()?;
        let clients = context.clients;

        let mysql = clients.mysql_pool();

        let id = input.id;
        let effect_name = input.name.clone();
        let repeat = input.repeat;
        let frames: Vec<LEDEffectFrame> = input
            .frames
            .iter()
            .map(|frame| LEDEffectFrame {
                leds: frame.leds.clone(),
                fade: frame.fade,
                start: frame.start,
            })
            .collect();

        // check if effect exists
        let led_effect = match sqlx::query!(
            r#"
                SELECT
                    LEDEffect.*,
                    Model.name as "model_name",
                    Part.name as "part_name"
                FROM LEDEffect
                INNER JOIN Model ON LEDEffect.model_id = Model.id
                INNER JOIN Part ON LEDEffect.part_id = Part.id
                WHERE LEDEffect.id = ?;
            "#,
            id
        )
        .fetch_one(mysql)
        .await
        {
            Ok(led_effect) => led_effect,
            Err(_) => {
                return Ok(LEDEffectResponse {
                    id: -1,
                    model_name: "".to_string(),
                    part_name: "".to_string(),
                    effect_name: "".to_string(),
                    repeat: 0,
                    effects: vec![],
                    ok: false,
                    msg: "effectName do not exist.".to_string(),
                })
            }
        };

        // check if effect is being edited by another user
        if let Ok(effect) = sqlx::query!(
            r#"
                SELECT * FROM EditingLEDEffect WHERE led_effect_id = ?;
            "#,
            id
        )
        .fetch_one(mysql)
        .await
        {
            if effect.user_id != context.user_id {
                return Err(GQLError::from(format!(
                    "This frame is being edited by user {}.",
                    effect.user_id
                )));
            }
        };

        // create array of all color ids from db
        let colors = sqlx::query!(
            r#"
                SELECT id FROM Color;
            "#
        )
        .fetch_all(mysql)
        .await?;

        // check if all color ids in frames are in db
        for frame in frames.iter() {
            for led in frame.leds.iter() {
                let mut found = false;
                for color in colors.iter() {
                    if led[0] == color.id {
                        found = true;
                        break;
                    }
                }
                if !found {
                    return Ok(LEDEffectResponse {
                        id: -1,
                        model_name: "".to_string(),
                        part_name: "".to_string(),
                        effect_name: "".to_string(),
                        repeat: 0,
                        effects: vec![],
                        ok: false,
                        msg: format!("Color Id {} not found", led[0]),
                    });
                }
            }
        }

        // update LEDEffect
        let _ = sqlx::query!(
            r#"
                UPDATE LEDEffect
                SET name = ?
                WHERE id = ?
            "#,
            &effect_name,
            id,
        )
        .execute(mysql)
        .await?;

        // update LEDEffectStates
        for frame in frames.iter() {
            for (i, led) in frame.leds.iter().enumerate() {
                let _ = sqlx::query!(
                    r#"
                        UPDATE LEDEffectState
                        SET color_id = ?, alpha = ?
                        WHERE position = ? AND effect_id = ?
                    "#,
                    led[0],
                    led[1],
                    i as i32,
                    id
                )
                .execute(mysql)
                .await?;
            }
        }

        let led_payload = LEDPayload {
            create_effects: Vec::new(),
            update_effects: vec![LEDEffectData {
                id,
                name: effect_name.clone(),
                model_name: led_effect.model_name.clone(),
                part_name: led_effect.part_name.clone(),
                repeat,
                frames: frames.clone(),
            }],
            delete_effects: Vec::new(),
        };

        Subscriptor::publish(led_payload);

        Ok(LEDEffectResponse {
            id,
            model_name: led_effect.model_name,
            part_name: led_effect.part_name,
            effect_name,
            repeat,
            effects: frames,
            ok: true,
            msg: "successfully edited LED effect".to_string(),
        })
    }

    #[graphql(name = "deleteLEDEffect")]
    async fn delete_led_effect(
        &self,
        ctx: &Context<'_>,
        id: i32,
    ) -> GQLResult<DeleteLEDEffectResponse> {
        let context = ctx.data::<UserContext>()?;
        let clients = context.clients;

        let mysql = clients.mysql_pool();

        // check if effect exists
        let led_effect = match sqlx::query!(
            r#"
                SELECT
                    LEDEffect.*,
                    Model.name as "model_name",
                    Part.name as "part_name"
                FROM LEDEffect
                INNER JOIN Model ON LEDEffect.model_id = Model.id
                INNER JOIN Part ON LEDEffect.part_id = Part.id
                WHERE LEDEffect.id = ?;
            "#,
            id
        )
        .fetch_one(mysql)
        .await
        {
            Ok(led_effect) => led_effect,
            Err(_) => {
                return Ok(DeleteLEDEffectResponse {
                    ok: false,
                    msg: format!("LEDEffect Id {} not found", id),
                })
            }
        };

        // check control data
        let control_frames = sqlx::query!(
            r#"
                SELECT frame_id FROM ControlData WHERE effect_id = ?;
            "#,
            id
        )
        .fetch_all(mysql)
        .await?;

        if !control_frames.is_empty() {
            let mut control_frames = control_frames
                .iter()
                .map(|frame| frame.frame_id)
                .collect::<Vec<i32>>();
            control_frames.sort();
            control_frames.dedup();

            return Ok(DeleteLEDEffectResponse {
                ok: false,
                msg: format!(
                    "LEDEffect Id {} is being used in control frames {:?}.",
                    id, control_frames
                ),
            });
        }

        // check if effect is being edited by another user
        if let Ok(effect) = sqlx::query!(
            r#"
                SELECT * FROM EditingLEDEffect WHERE led_effect_id = ?;
            "#,
            id
        )
        .fetch_one(mysql)
        .await
        {
            if effect.user_id != context.user_id {
                return Err(GQLError::from(format!(
                    "This frame is being edited by user {}.",
                    effect.user_id
                )));
            }
        };

        // delete from LEDEffectStates
        let _ = sqlx::query!(
            r#"
                DELETE FROM LEDEffectState
                WHERE effect_id = ?
            "#,
            id
        )
        .execute(mysql)
        .await?;

        // delete from LEDEffect
        let _ = sqlx::query!(
            r#"
                DELETE FROM LEDEffect
                WHERE id = ?
            "#,
            id
        )
        .execute(mysql)
        .await?;

        // publish to subscribers
        let led_payload = LEDPayload {
            create_effects: Vec::new(),
            update_effects: Vec::new(),
            delete_effects: vec![LEDEffectData {
                id,
                name: "".to_string(),
                model_name: led_effect.model_name.clone(),
                part_name: led_effect.part_name.clone(),
                repeat: 0,
                frames: vec![],
            }],
        };

        Subscriptor::publish(led_payload);

        Ok(DeleteLEDEffectResponse {
            ok: true,
            msg: "successfully deleted LED effect".to_string(),
        })
    }
}
